### DISCUSSION TITLE
Why tsp install?

### DISCUSSION THREAD

**Discussion by BlairCurrey** [microsoft/typespec#3078] [Q&A] [answered]
I'm following the guide in the readme which says: 

Install the dependencies of main.tsp:
    
    tsp install

As opposed to `npm install`, or what I would prefer, `pnpm install` as I'm in a pnpm workspace.  It looks like the `tsp install` command does `npm install` under the hood but presumably it does more than that (otherwise why have it)?

Doing `pnpm install` instead seems to work (can build the openapi spec as expected). Is there some case or way in which this doesnt work? 

To try and answer my question, I think it's only doing `npm install`. This is the conclusion I'm coming to from looking at the `installTypeSpecDependencies` [implementation](https://github.com/microsoft/typespec/blob/4e1d30ddce3abcfbc8e5c0bef777d16ed68eef1a/packages/compiler/src/core/install.ts#L12) and [usage](https://github.com/microsoft/typespec/blob/4e1d30ddce3abcfbc8e5c0bef777d16ed68eef1a/packages/compiler/src/core/cli/cli.ts#L218)

If this is the case, and installing with pnpm, yarn, etc. is fine, I think the instructions should be to use `npm install`. That implies you can just install normally via your package manager. I recognize `TypeSpec` can be used outside the js ecosystem but `npm` can be assumed because the instructions also include a step for installing it.

**Comment by timotheeguerin** [comment, marked as answer]
It does only do `npm install` today, only reason it added was to hide the fact that you needed yet another command from people not familiar with the node ecosystem. 
You are totally able to use any package manager that produce node compatible `node_modules`, we use `pnpm` as well internally in our repos.
But I do agree that we should make it clearer that other package manager can be used.

**Reply by BlairCurrey** [reply]
Thank you for clarifying.

> only reason it added was to hide the fact that you needed yet another command from people not familiar with the node ecosystem.

Ya... that make sense. Although an `npm` command is already used in the first step. So it's kinda just partially hiding the package management. I wonder how much that helps people unfamiliar with node package management.

> But I do agree that we should make it clearer that other package manager can be used.

That does address my main concern. Coming from a node perspective it just looks a bit curious. It's useful to know that `tsp install` isn't doing anything special under the hood.

**Reply by attilah** [reply]
Also as a suggestion it could use the common package manager detection algos to use the one that is used for the project. Then it would be opaque command for the ones who familiar with it. What we don't want is to have a new lock file for the dependencies :-)

**Reply by Downchuck** [reply]
Would be nice to see yarn pnp supported

**Reply by timotheeguerin** [reply]
@Downchuck in what way would you like to see that? We just have tsp install that map to npm install but you are free to use whatever package manager you want

**Reply by Downchuck** [reply]
I'll give it a shot as I may have misrun it -- I was thinking "yarn tsp compile" much as one can run "yarn node" - but I haven't tried the "yarn dlx" path which may be OK.

I did have a bit of a struggle in the Visual Studio Code Typespec plugin as it expected plugins in the repo in node_modules it seemed.

I'll post a new thread as this comment is clearly more about using tsp from a yarn pnp repo than it is about "tsp install" itself. Using yarn with pnpm mode (linked node modules) works as expected with tsp.
